instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/HttpResponseReceiverInstrumenter.java

```diff
@@ -0,0 +1,182 @@
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import reactor.core.publisher.Mono;
+import reactor.netty.Connection;
+import reactor.netty.ConnectionObserver;
+import reactor.netty.http.client.HttpClient;
+import reactor.netty.http.client.HttpClientConfig;
+import reactor.netty.http.client.HttpClientRequest;
+import reactor.netty.http.client.HttpClientResponse;
+import reactor.netty.http.client.HttpClientState;
+
+public final class HttpResponseReceiverInstrumenter {
+
+  // this method adds several stateful listeners that execute the instrumenter lifecycle during HTTP
+  // request processing
+  // it should be used just before one of the response*() methods is called - at this point the HTTP
+  // request is no longer modifiable by the user
+  @Nullable
+  public static HttpClient.ResponseReceiver<?> instrument(HttpClient.ResponseReceiver<?> receiver) {
+    // receiver should always be an HttpClientFinalizer, which both extends HttpClient and
+    // implements ResponseReceiver
+    if (receiver instanceof HttpClient) {
+      HttpClient client = (HttpClient) receiver;
+      HttpClientConfig config = client.configuration();
+
+      ContextHolder contextHolder = new ContextHolder();
+
+      HttpClient modified =
+          client
+              .mapConnect(new StartOperation(contextHolder, config))
+              .doOnRequest(new PropagateContext(contextHolder))
+              .doOnRequestError(new EndOperationWithError(contextHolder, config))
+              .observe(new EndOperation(contextHolder, config));
+
+      // modified should always be an HttpClientFinalizer too
+      if (modified instanceof HttpClient.ResponseReceiver) {
+        return (HttpClient.ResponseReceiver<?>) modified;
+      }
+    }
+
+    return null;
+  }
+
+  static final class ContextHolder {
+    volatile Context parentContext;
+    volatile Context context;
+  }
+
+  static final class StartOperation
+      implements Function<Mono<? extends Connection>, Mono<? extends Connection>> {
+
+    private final ContextHolder contextHolder;
+    private final HttpClientConfig config;
+
+    StartOperation(ContextHolder contextHolder, HttpClientConfig config) {
+      this.contextHolder = contextHolder;
+      this.config = config;
+    }
+
+    @Override
+    public Mono<? extends Connection> apply(Mono<? extends Connection> mono) {
+      return Mono.defer(
+          () -> {
+            Context parentContext = Context.current();
+            contextHolder.parentContext = parentContext;
+            if (!instrumenter().shouldStart(parentContext, config)) {
+              return mono.contextWrite(ctx -> ctx.put(CLIENT_PARENT_CONTEXT_KEY, parentContext));
+            }
+
+            Context context = instrumenter().start(parentContext, config);
+            contextHolder.context = context;
+            return ContextPropagationOperator.runWithContext(mono, context)
```

Comment:
nice! it looks like pure context API is still useful even without extra convenience we discussed to start spans

Reply:
Yes, it turned out to be super useful this time 🚀